textalyze.rb

```diff
@@ -1,66 +1,101 @@
-# This is the base code for v0.1 of our Ruby text analyzer.
-# Visit https://github.com/codeunion/text-analysis/wiki to see what to do.
-#
-# Send an email to your cohort mailing list if you have any questions
-# or you're stuck!  These comments are here to help you, but please delete them
-# as you go along. You wouldn't normally have such heavily-commented code.
+require_relative './max'
 
-# Method name: item_counts
-# Input:   An arbitrary array
-#
-# Returns: A hash where every item is a key whose value is the number of times
-#          that item appears in the array
-#
-# Prints:  Nothing
+if ARGV.size == 0
+  puts "You didn't tell us what file to read. Try this commmand:"
+  puts
+  puts "  ruby textalyze.rb random_file.txt"
+  exit # Exits the program
+end
 
-# Here are some examples:
-#     item_counts(["I", "am", "Sam", "I", "am"])
-#       # => {"I" => 2, "am" => 2, "Sam" => 1}
-#
-#     item_counts([10, 20, 10, 20, 20])
-#       # => {10 => 2, 20 => 3}
-#
-# In short, item_counts(array) tells us how many times each item appears
-# in the input array.
+file_name = ARGV[0] #first argument
+file_contents = File.read(file_name) # Files will have to be in same directory
+# as textalyze.rb - simimlar to v0.4.
 
 def item_counts(array)
   counts = {} # Initialize counts to an empty Hash
+  array.each do |key|
+    if counts[key] # If this key exists
+      counts[key] += 1 # Increment the key
+    else # If this key doesn't exist
+      counts[key] = 1 # Create the key, set value to 1
+    end
+  end
+  # print_counts(counts)
+  counts
+end
 
-  array.each do |item|
-    # Add code here to modify the "counts" hash accordingly
-    # You'll need to handle two cases:
-    #   1. The first time we've seen a particular item in the array
-    #   2. The second-or-later time we've seen a particular item in the array
+def print_counts(counts_hash)
+  puts
+  counts_hash.each do |key, value|
+    puts "#{key}   #{value}"
   end
+  puts
+end
+
+def string_to_array(string) # returns array of all characters in the string
+  sanitize(string)
+  string.split(//)
+end
+
+def sanitize(string)
+  string.downcase!
```

Comment:
As a rule, avoid the destructive versions of `downcase` and `gsub` and other methods.  By convention, they're denoted with an `!` at the end of the method name.  The `!` itself doesn't have any special meaning to Ruby — it's just another character as far as its concerns — but it signals to the programmer that this method is destructive.

"Destructive" means that the method changes the value of the thing it's being called on vs. returning a new copy with the appropriate changes.  In this instance, `sanitize` will permanently alter any `String` we pass it.

I'd write the method like this:

```ruby
def sanitize(string)
  string.downcase.gsub(/\W+/, "")
end
```